staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go

```diff
@@ -0,0 +1,394 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	restful "github.com/emicklei/go-restful"
+	"github.com/go-openapi/spec"
+
+	v1 "k8s.io/api/autoscaling/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	metav1beta1 "k8s.io/apimachinery/pkg/apis/meta/v1beta1"
+	"k8s.io/apiserver/pkg/endpoints"
+	"k8s.io/apiserver/pkg/endpoints/openapi"
+	"k8s.io/klog"
+	openapibuilder "k8s.io/kube-openapi/pkg/builder"
+	"k8s.io/kube-openapi/pkg/common"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	generatedopenapi "k8s.io/kubernetes/pkg/generated/openapi"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+)
+
+const (
+	// objectMetaType = "k8s.io/apimachinery/pkg/apis/meta/v1.ObjectMeta"
+	objectMetaSchemaRef = "#/definitions/io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta"
+	listMetaType        = "k8s.io/apimachinery/pkg/apis/meta/v1.ListMeta"
+	typeMetaType        = "k8s.io/apimachinery/pkg/apis/meta/v1.TypeMeta"
+
+	definitionPrefix = "#/definitions/"
+)
+
+var swaggerPartialObjectMetadataDescriptions = metav1beta1.PartialObjectMetadata{}.SwaggerDoc()
+
+// BuildSwagger builds swagger and calculates etag for given crd in given version
+func BuildSwagger(crd *apiextensions.CustomResourceDefinition, version string) (*spec.Swagger, string, error) {
+	schema := &spec.Schema{}
+	s, err := getSchemaForVersion(crd, version)
+	if err != nil {
+		return nil, "", err
+	}
+	if s != nil && s.OpenAPIV3Schema != nil {
+		schema, err = ConvertJSONSchemaPropsToOpenAPIv2Schema(s.OpenAPIV3Schema)
+		if err != nil {
+			return nil, "", err
+		}
+	}
+	b := newBuilder(crd, version, schema)
+
+	// Sample response types for building web service
+	sample := sampleType{}
+	sampleList := sampleListType{}
+	status := metav1.Status{}
+	patch := metav1.Patch{}
+	scale := v1.Scale{}
+
+	routes := make([]*restful.RouteBuilder, 0)
+	root := fmt.Sprintf("/apis/%s/%s/%s", b.group, b.version, b.plural)
+	if b.namespaced {
+		routes = append(routes, b.buildRoute(root, "", "GET", "list", sampleList).
+			Operation("list"+b.kind+"ForAllNamespaces"))
+		root = fmt.Sprintf("/apis/%s/%s/namespaces/{namespaces}/%s", b.group, b.version, b.plural)
+	}
+	routes = append(routes, b.buildRoute(root, "", "GET", "list", sampleList))
```

Comment:
I am surprised to see all this logic here which should be generic for every resource, not only CRDs. Where is it copied from?

Reply:
The logic comes from https://github.com/kubernetes/kubernetes/blob/8b98e802eddb9f478ff7d991a2f72f60c165388a/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L166 It's quite complex for built-in types

Reply:
IMO it is a no-go to move forward with the PR as long as this whole code is duplicated.